src/main/java/io/vertx/core/shareddata/impl/LocalAsyncLocks.java

```diff
@@ -49,12 +47,30 @@
       timerId = timeout != Long.MAX_VALUE ? context.owner().setTimer(timeout, tid -> timeout()) : null;
     }
 
-    boolean isWaiting() {
-      return status.get() == Status.WAITING;
-    }
-
     void timeout() {
       if (status.compareAndSet(Status.WAITING, Status.TIMED_OUT)) {
+        // Cleanup
+        waitersMap.compute(lockName, (s, list) -> {
+          if (list == null) {
+            // Already removed by nextWaiter
+            return null;
+          } else {
+            int idx = list.indexOf(LockWaiter.this);
+            if (idx == -1) {
+              // Already removed by nextWaiter
+              return null;
+            }
+            int size = list.size();
+            ArrayList<LockWaiter> n = new ArrayList<>(size - 1);
```

Comment:
Could we be more aggressive and check if we can return `null` when there's no other waiter or a `SingletonList` when there's just one?

Reply:
yes we need to consider this case as it could be an actual one and leave an empty list in the map, I'll try to address this with a test